linalg/hypre.cpp

```diff
@@ -1713,6 +1713,292 @@ HypreParMatrix * RAP(const HypreParMatrix * Rt, const HypreParMatrix *A,
    return new HypreParMatrix(rap);
 }
 
+// Helper function for HypreParMatrixFromBlocks
+void GatherBlockOffsetData(MPI_Comm comm, const int rank, const int nprocs,
+                           const int num_loc, Array<int> &offsets,
+                           std::vector<int> &all_num_loc, const int numBlocks,
+                           std::vector<std::vector<int>> &blockProcOffsets,
+                           std::vector<int> &procOffsets,
+                           std::vector<std::vector<int>> &procBlockOffsets,
+                           int &firstLocal, int &globalNum)
+{
+   std::vector<std::vector<int>> all_block_num_loc(numBlocks);
+
+   MPI_Allgather(&num_loc, 1, MPI_INT, all_num_loc.data(), 1, MPI_INT, comm);
+
+   for (int j = 0; j < numBlocks; ++j)
+   {
+      all_block_num_loc[j].resize(nprocs);
```

Comment:
For our purposes this is probably fine, but one thing Hypre has worked hard to do is not have arrays of size `nprocs` anywhere so that it can scale to arbitrary processor counts (this also implies replacing `MPI_Allgather` with point-to-point communcations). Maybe add a comment somewhere noting that this function is not truly scalable to extreme processor counts?

Reply:
I wish I had the problem of too many processors, but in any case this is a good point that needs to be documented.